### PR TITLE
Audio source picker when bluetooth device is present

### PR DESCRIPTION
**Pull request by michaelkirk**
When no bluetooth is detected, the call screen is unchanged.

However when we *do* detect a bluetooth device, we change the "toggle speaker phone" button to be a "pick audio source" button.

img_0960

The "pick audio source" action sheet shows which are available, and marks the currently active source. I'd like to add icons to the action sheet items to make it easier to distinguish between devices, but wanted to get a PR open.

img_0961

With bluetooth connected, the audio picker button needs to be visible with the video call view. This unfortunately obscures more of the video. Any ideas?

img_0964

### CODE REVIEW

**Comment by charlesmchen** [comment]
> With bluetooth connected, the audio picker button needs to be visible with the video call view. This unfortunately obscures more of the video. Any ideas?

My personal inclination is to use a "T" configuration in that case, so the top row is (mute, audio device, video) and the second row is (hang up). Yes, that'll _worsen_ the issue since it moves buttons towards the center of the screen but it'll be a safer (fewer accidental hangups) and be consistent with other modes of that view.  

I think we previously discussed using mostly-transparent buttons when remote video is active. Hmm.

File: Signal/Images.xcassets/ic_speaker_bluetooth_inactive_audio_mode.imageset/Contents.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      "scale" : "2x"
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "3x"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
```
Comment (charlesmchen) [inline Signal/Images.xcassets/ic_speaker_bluetooth_inactive_audio_mode.imageset/Contents.json]: Where'd you get this asset?  Also, don't forget to update https://github.com/WhisperSystems/Signal-iOS-Design.  I've been continuing to push changes to a branch https://github.com/WhisperSystems/Signal-iOS-Design/tree/charlesmchen/addIcons but I'm thinking of just making changes in master.

File: Signal/src/ViewControllers/CallViewController.swift [hunk 2]
```diff
@@ -86,18 +86,53 @@ class CallViewController: UIViewController, CallObserver, CallServiceObserver, R
     var settingsNagView: UIView!
     var settingsNagDescriptionLabel: UILabel!
 
+    // MARK: Audio Source
+
+    var hasAlternateAudioSources: Bool {
+        Logger.info("\(TAG) available audio routes count: \(allAudioSources.count)")
+        // internal mic and speakerphone will be the first two, any more than one indicates e.g. an attached bluetooth device.
+
+        // TODO is this sufficient? Are their devices w/ bluetooth but no external speaker? e.g. ipod?
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CallViewController.swift]: I think this is fine. iPod Touches have speakers.

File: Signal/src/ViewControllers/CallViewController.swift [hunk 3]
```diff
@@ -305,12 +345,53 @@ class CallViewController: UIViewController, CallObserver, CallServiceObserver, R
...
+        ])
+    }
+
+    func didChangeAudioSession() {
+        AssertIsOnMainThread()
+
+        // Which sources are available depends on the state of your Session.
+        // When the audio session is not yet in PlayAndRecord none are available
+        // Then if we're in speakerphone, bluetooth isn't available. 
+        // So we acrew all possible audio sources in a set, and that list lives as longs as the CallViewController
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CallViewController.swift]: s/acrew/accrue/

Comment (charlesmchen) [inline Signal/src/ViewControllers/CallViewController.swift]: s/longs/long/
s/CallViewController/CallViewController./

File: Signal/src/ViewControllers/CallViewController.swift [hunk 4]
```diff
@@ -305,12 +345,53 @@ class CallViewController: UIViewController, CallObserver, CallServiceObserver, R
...
+        // The downside of this is that if you e.g. unpair your bluetooth mid call, it will still appear as an option
+        // until your next call.
+        // FIXME: There's got to be a better way, but this is where I landed after a bit of work, and seems to work
+        // pretty well in practrice.
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CallViewController.swift]: s/practrice/practice/

File: Signal/src/ViewControllers/CallViewController.swift [hunk 5]
```diff
@@ -685,6 +766,32 @@ class CallViewController: UIViewController, CallObserver, CallServiceObserver, R
...
+            // No bluetooth audio detected
+
+            audioSourceButton.isSelected = call.isSpeakerphoneEnabled
+            audioSourceButton.setImage(#imageLiteral(resourceName: "audio-call-speaker-inactive"), for: .normal)
+            audioSourceButton.setImage(#imageLiteral(resourceName: "audio-call-speaker-active"), for: .selected)
+
+            // If there's no bluetooth, we always use speakerphone, so no need for
+            // a button, giving more screen back for the video.
+            audioSourceButton.isHidden = hasLocalVideo
+        }
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CallViewController.swift]: this method is so long...

File: Signal/src/call/CallAudioService.swift [hunk 6]
```diff
@@ -5,6 +5,71 @@
...
+                         image: #imageLiteral(resourceName: "button_phone_white"), //TODO
+                         isBuiltInSpeaker: true)
+    }
+
+    // MARK: Hashable
+
+    static func ==(lhs: AudioSource, rhs: AudioSource) -> Bool {
+        // Simply comparing the `portDescription` vs the `portDescription.uid`
+        // caused multiple instances of the built in mic to turn up in a set.
+        if lhs.isBuiltInSpeaker && rhs.isBuiltInSpeaker {
```
Comment (charlesmchen) [inline Signal/src/call/CallAudioService.swift]: The comment refers to "built-in mic" but the code refers to "built-in speaker."

File: Signal/src/call/CallAudioService.swift [hunk 7]
```diff
@@ -74,10 +139,19 @@ import AVFoundation
         Logger.verbose("\(TAG) in \(#function) is no-op")
     }
 
-    internal func speakerphoneDidChange(call: SignalCall, isEnabled: Bool) {
+    internal func audioSourceDidChange(call: SignalCall, audioSource: AudioSource?) {
         AssertIsOnMainThread()
 
         ensureProperAudioSession(call: call)
+
+        // It's importent to set preferred input *after* ensuring properAudioSession
```
Comment (charlesmchen) [inline Signal/src/call/CallAudioService.swift]: s/importent/important/

File: Signal/Images.xcassets/ic_speaker_bluetooth_inactive_audio_mode.imageset/Contents.json (hunk 1 above)
Comment (michaelkirk) [inline Signal/Images.xcassets/ic_speaker_bluetooth_inactive_audio_mode.imageset/Contents.json]: > Where'd you get this asset?

I made it in an attempt to match our existing style, while matching platform conventions. In particular, I didn't want the speaker portion of the button to be positioned differently in the speakerphone vs. bluetooth.

>  Also, don't forget to update ...
Yup, I've got them in Signal-iOS-Design locally, waiting until this is merged.

File: Signal/src/ViewControllers/CallViewController.swift (hunk 3 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/CallViewController.swift]: :set spell

Thanks =)

**Pull request merged**